### PR TITLE
[GLUTEN-6053][CH] Move collect native metrics from last hasNext to close and cancel

### PR DESCRIPTION
**Pull request by lwz9103**
## What changes were proposed in this pull request?

(Fixes: \#6053)

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by lwz9103** [comment]
Run Gluten Clickhouse CI

File: backends-clickhouse/src/main/scala/org/apache/gluten/backendsapi/clickhouse/CHIteratorApi.scala [hunk 1]
```diff
@@ -43,6 +43,7 @@ import java.util.{ArrayList => JArrayList, HashMap => JHashMap, Map => JMap}
 
 import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
+import scala.language.reflectiveCalls
```
Comment (zzcclp) [inline backends-clickhouse/src/main/scala/org/apache/gluten/backendsapi/clickhouse/CHIteratorApi.scala:46]: remove ?

Comment (lwz9103) [reply backends-clickhouse/src/main/scala/org/apache/gluten/backendsapi/clickhouse/CHIteratorApi.scala:46]: remove this will cause compile error

Comment (baibaichen) [reply backends-clickhouse/src/main/scala/org/apache/gluten/backendsapi/clickhouse/CHIteratorApi.scala:46]: `genFirstStageIterator`  里的 `val iter = new Iterator[Any]` 和 `genFinalStageIterator` 里的  `val resIter = new Iterator[ColumnarBatch]` 基本重复，主要区别就是是否 `updateInputMetrics(inputMetrics)`，如果这里新建一个 `CollectMetricItertor`，一方面可以去除重复代码，另一方面应该可以不用 `import scala.language.reflectiveCalls`，下图是我改了变量名之后的 diff

img_v3_02bs_c2b29209-2de7-49e4-a421-596f384bd5cg

**Review by baibaichen — APPROVED** [review]

**Pull request merged**